internal/rpmmd/repository.go

```diff
@@ -534,6 +543,155 @@ func (r *rpmmdImpl) Depsolve(packageSet PackageSet, repos []RepoConfig, modulePl
 	return dependencies, reply.Checksums, err
 }
 
+// ChainPackageSets constructs an array of `ChainPackageSet` based on the provided
+// arguments. The provided `packageSets` map is transformed into an array based
+// on the order of package set names passed in `packageSetsChain`. Repositories
+// provided in `repos` are used for every transaction, while repositories from
+// `packageSetsRepos` are used only for package sets with the respective name.
+//
+// The function returns a `ChainPackageSet` slice, which members are referencing
+// repositories from the returned `RepoConfig` by their index. The returned
+// `RepoConfig` slice is specific to the requested `packageSetsChain`.
+//
+// NOTE: Due to implementation limitations of DNF and dnf-json, each package set
+// in the chain must use all of the repositories used by its predecessor.
+// An error is returned if this requirement is not met.
+func chainPackageSets(packageSetsChain []string, packageSets map[string]PackageSet, repos []RepoConfig, packageSetsRepos map[string][]RepoConfig) ([]chainPackageSet, []RepoConfig, error) {
+	transactions := make([]chainPackageSet, 0, len(packageSetsChain))
+	transactionsRepos := make([]RepoConfig, 0, len(repos))
+
+	transactionsRepos = append(transactionsRepos, repos...)
+
+	// These repo IDs will be used for every transaction
+	baseRepoIDs := make([]int, 0, len(repos))
+	for idx := range repos {
+		baseRepoIDs = append(baseRepoIDs, idx)
+	}
+
+	for transactionIdx, pkgSetName := range packageSetsChain {
+		pkgSet, ok := packageSets[pkgSetName]
+		if !ok {
+			return nil, nil, fmt.Errorf("package set %q requested in the 'packageSetsChain' does not exist in provided 'packageSets'", pkgSetName)
+		}
+
+		transaction := chainPackageSet{
+			PackageSet: pkgSet,
+			Repos:      baseRepoIDs, // Due to its capacity, the slice will be copied if any repo is appended
+		}
+
+		// Add any package-set-specific repos to the list of transaction repos
+		if pkgSetRepos, ok := packageSetsRepos[pkgSetName]; ok {
+			for _, pkgSetRepo := range pkgSetRepos {
+				// Check if the repo has been already used by a transaction
+				// and if yes, just use its ID. Skip the "base" repos.
+				pkgSetRepoID := -1
+				for idx := len(repos); idx < len(transactionsRepos); idx++ {
+					transactionRepo := transactionsRepos[idx]
+					if reflect.DeepEqual(pkgSetRepo, transactionRepo) {
```

Comment:
I always have mixed feelings about `reflect` but it's probably fine here.

Reply:
The other option would be to implement a logic which would do the deep equality check. Given the documentation for `reflect.DeepEqual()` I decided that it basically does what I needed in this case and that it should be safe to use it to compare two `RepoConfig` instances.

But I'm happy to rework this if you have any suggestions 😉

Reply:
Yeah, agreed, it's actually very sane (if you have sane structures).